File: ruby/food-chain/food_chain.rb
# Food Chain Song
class FoodChainSong
  ANIMALS = {
    1 => :fly,
    2 => :spider,
    3 => :bird,
    4 => :cat,
    5 => :dog,
    6 => :goat,
    7 => :cow,
    8 => :horse
  }.freeze

  SECOND_LINES = {
    fly: '',
    spider: "It wriggled and jiggled and tickled inside her.\n",
    bird: "How absurd to swallow a bird!\n",
    cat: "Imagine that, to swallow a cat!\n",
    dog: "What a hog, to swallow a dog!\n",
    goat: "Just opened her throat and swallowed a goat!\n",
    cow: "I don't know how she swallowed a cow!\n",
    horse: "She's dead, of course!\n"
  }.freeze

  def verses(start = 1, stop = 8)
    (start..stop).map { |verse_num| verse(verse_num) }.join("\n") + "\n"
  end
  alias sing verses

  def verse(verse_num)
    animal = ANIMALS[verse_num]
    return head(animal) if animal == :horse
    head(animal) + middle_lines(verse_num) + tail
  end

  def head(animal)
    ["I know an old lady who swallowed a #{animal}.",
     SECOND_LINES[animal]].join("\n")
  end

  def middle_lines(verse_num)
    verse_num.downto(1).map do |curr|
      next if curr == 1
      animal1 = ANIMALS[curr]
      animal2 = ANIMALS[curr - 1]
      middle_line(animal1, animal2)
    end.join("\n")
  end

  def middle_line(animal1, animal2)
    "She swallowed the #{animal1} to catch the #{animal2}#{spider(animal2)}."
  end

  def spider(animal2)
    ' that wriggled and jiggled and tickled inside her' if animal2 == :spider
  end

  def tail
    "I don't know why she swallowed the fly. Perhaps she'll die.\n"
  end
end

File: javascript/food-chain/food-chain.js
// Copy solution from:
// https://github.com/breethomas/exercisms/blob/master/javascript/food-chain/food-chain.js

var Song = function(lyrics) {
  this.verseArray = lyrics.split("\n\n");
};

Song.prototype.verse = function(start) {
  return this.verseArray[start - 1] + "\n";
};

Song.prototype.verses = function(start, end) {
  var sliced = this.verseArray.slice(start - 1, end);
  return sliced.join("\n\n") + "\n\n";
};

Song.prototype.sing = function() {
  var end = this.verseArray.length;
  return this.verses(1, end);
};

Song.oldLadyTune = function() {
  return new Song(oldLadyTune);
};

var oldLadyTune =
  "\
I know an old lady who swallowed a fly.\n\
I don't know why she swallowed the fly. Perhaps she'll die.\n\
\n\
I know an old lady who swallowed a spider.\n\
It wriggled and jiggled and tickled inside her.\n\
She swallowed the spider to catch the fly.\n\
I don't know why she swallowed the fly. Perhaps she'll die.\n\
\n\
I know an old lady who swallowed a bird.\n\
How absurd to swallow a bird!\n\
She swallowed the bird to catch the spider that wriggled and jiggled and tickled inside her.\n\
She swallowed the spider to catch the fly.\n\
I don't know why she swallowed the fly. Perhaps she'll die.\n\
\n\
I know an old lady who swallowed a cat.\n\
Imagine that, to swallow a cat!\n\
She swallowed the cat to catch the bird.\n\
She swallowed the bird to catch the spider that wriggled and jiggled and tickled inside her.\n\
She swallowed the spider to catch the fly.\n\
I don't know why she swallowed the fly. Perhaps she'll die.\n\
\n\
I know an old lady who swallowed a dog.\n\
What a hog, to swallow a dog!\n\
She swallowed the dog to catch the cat.\n\
She swallowed the cat to catch the bird.\n\
She swallowed the bird to catch the spider that wriggled and jiggled and tickled inside her.\n\
She swallowed the spider to catch the fly.\n\
I don't know why she swallowed the fly. Perhaps she'll die.\n\
\n\
I know an old lady who swallowed a goat.\n\
Just opened her throat and swallowed a goat!\n\
She swallowed the goat to catch the dog.\n\
She swallowed the dog to catch the cat.\n\
She swallowed the cat to catch the bird.\n\
She swallowed the bird to catch the spider that wriggled and jiggled and tickled inside her.\n\
She swallowed the spider to catch the fly.\n\
I don't know why she swallowed the fly. Perhaps she'll die.\n\
\n\
I know an old lady who swallowed a cow.\n\
I don't know how she swallowed a cow!\n\
She swallowed the cow to catch the goat.\n\
She swallowed the goat to catch the dog.\n\
She swallowed the dog to catch the cat.\n\
She swallowed the cat to catch the bird.\n\
She swallowed the bird to catch the spider that wriggled and jiggled and tickled inside her.\n\
She swallowed the spider to catch the fly.\n\
I don't know why she swallowed the fly. Perhaps she'll die.\n\
\n\
I know an old lady who swallowed a horse.\n\
She's dead, of course!";

module.exports = Song.oldLadyTune();
